### PR TITLE
Topic/cl hierarchical

### PR DESCRIPTION
**Pull request by vspetrov**
## What
Introduces base implementation of CL/HIER (CL that implements hierarchical collective algorithms) and its integration into CORE.

## Why ?
CL/HIER will the framework for hierarchy based colls: small message latency bounded (barrier, allreduce, reduce, bcast); large msg pipelined bcast; hybrid large message allreduce; etc.

## How ?
CL/HIER builds collective schedules by combining process subgroups (ucc_sbgp_t) and TL primitives on the sbgps. 
Integration of multiple CLs (basic and hier) is made in the core via the same score_map mechanism.
This PR does not introduce any particular algorithm in CL/HIER, only brings infra.

### CODE REVIEW

File: src/components/cl/hier/cl_hier_context.c [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+{
+    const ucc_cl_context_config_t *cl_config =
+        ucc_derived_of(config, ucc_cl_context_config_t);
+    ucc_config_names_array_t *tls = &cl_config->cl_lib->tls;
+    ucc_status_t              status;
+    int                       i;
+
+    UCC_CLASS_CALL_SUPER_INIT(ucc_cl_context_t, cl_config->cl_lib,
+                              params->context);
+    if (tls->count == 1 && !strcmp(tls->names[0], "all")) {
```
Comment (Sergei-Lebedev) [inline src/components/cl/hier/cl_hier_context.c:22]: so we have UCC_CL_HIER_TLS, UCC_CL_HIER_NODE_SBGP_TLS and UCC_CL_HIER_NET_SBGP_TLS, what is the relationship between them? if for example HIER_TLS=ucp and NET_SBGP_TLS=sharp, NODE_SBGP_TLS=ucp what TLs will be used?

File: src/components/cl/hier/cl_hier_team.c [hunk 2]
```diff
@@ -0,0 +1,305 @@
...
+    n_sbgp_teams = 0;
+    for (i = 0; i < UCC_HIER_SBGP_LAST; i++) {
+        hs        = &self->sbgps[i];
+        hs->score = NULL;
+        if (hs->state == UCC_HIER_SBGP_ENABLED) {
+            hs->sbgp =
+                ucc_team_topo_get_sbgp(params->team->topo, hs->sbgp_type);
+            if (hs->sbgp->status != UCC_SBGP_ENABLED) {
+                /* SBGP of that type either not exists or the calling process
+                   is not part of part */
```
Comment (Sergei-Lebedev) [inline src/components/cl/hier/cl_hier_team.c:56]: typo "is not part of" ?

File: src/components/cl/hier/cl_hier_lib.c [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+    memset(&tl_attr, 0, sizeof(tl_attr));
+    status = tl_iface->lib.get_attr(NULL, &tl_attr.super);
+    if (UCC_OK != status) {
+        cl_error(lib, "failed to query tl %s lib attributes",
+                 tl_iface->super.name);
+        return status;
+    }
+    attr->super.attr.thread_mode =
+        ucc_min(attr->super.attr.thread_mode, tl_attr.super.attr.thread_mode);
+    attr->super.attr.coll_types |= tl_attr.super.attr.coll_types;
```
[the remark this answers is not in the document]
Comment (vspetrov) [reply src/components/cl/hier/cl_hier_lib.c:64]: no, copy-paste bug

File: src/components/cl/hier/Makefile.am [hunk 4]
```diff
@@ -0,0 +1,20 @@
+#
+# Copyright (C) Mellanox Technologies Ltd. 2020.  ALL RIGHTS RESERVED.
```
Comment (manjugv) [inline src/components/cl/hier/Makefile.am:2]: 2020-2021

File: src/components/cl/hier/cl_hier.c [hunk 5]
```diff
@@ -0,0 +1,55 @@
+/**
+ * Copyright (C) Mellanox Technologies Ltd. 2020.  ALL RIGHTS RESERVED.
```
Comment (manjugv) [inline src/components/cl/hier/cl_hier.c:2]: 2020-2021

File: src/components/cl/hier/cl_hier.h [hunk 6]
```diff
@@ -0,0 +1,102 @@
+/**
+ * Copyright (C) Mellanox Technologies Ltd. 2020.  ALL RIGHTS RESERVED.
```
Comment (manjugv) [inline src/components/cl/hier/cl_hier.h:2]: 2020-2021

File: src/components/cl/hier/cl_hier_context.c [hunk 7]
```diff
@@ -0,0 +1,75 @@
+/**
+ * Copyright (C) Mellanox Technologies Ltd. 2020.  ALL RIGHTS RESERVED.
```
Comment (manjugv) [inline src/components/cl/hier/cl_hier_context.c:2]: 2020-2021

File: src/components/cl/hier/cl_hier_lib.c [hunk 8]
```diff
@@ -0,0 +1,111 @@
+/**
+ * Copyright (C) Mellanox Technologies Ltd. 2020.  ALL RIGHTS RESERVED.
```
Comment (manjugv) [inline src/components/cl/hier/cl_hier_lib.c:2]: 2020-2021

File: src/coll_score/ucc_coll_score.c [hunk 9]
```diff
@@ -231,8 +231,8 @@ static ucc_status_t ucc_msg_range_add_fallback(const ucc_msg_range_t *in,
         }                                                                      \
     } while (0)
 
-static ucc_status_t ucc_score_list_dup(ucc_list_link_t *src,
-                                       ucc_list_link_t *dst)
+static ucc_status_t ucc_score_list_dup(const ucc_list_link_t *src,
+                                       ucc_list_link_t *      dst)
```
Comment (manjugv) [inline src/coll_score/ucc_coll_score.c:235]: clang formatting is inconsistent?

File: src/components/cl/hier/cl_hier_team.c [hunk 10]
```diff
@@ -0,0 +1,305 @@
...
+/* The function below must enable/disable those hier sbgps that will be
+   used to construct hierarchical schedules.
+   Currently just enable two sbgps as example and for testing purposes.
+   Next step is to enable sbgps based on the requested hierarchical algs. */
+static void ucc_cl_hier_enable_sbgps(ucc_cl_hier_team_t *team)
+{
+    SBGP_SET(team, NODE, ENABLED);
+    SBGP_SET(team, NET, ENABLED);
+}
+
```
Comment (manjugv) [inline src/components/cl/hier/cl_hier_team.c:25]: Do we expect two CLs to be enabled at the same time? Say, I want to do one collective with hierarchical team and another with basic team?

File: src/components/cl/hier/cl_hier_team.c [hunk 11]
```diff
@@ -0,0 +1,305 @@
...
+                                       tls->names[j], &hs->tl_ctxs[hs->n_tls]);
+                if (UCC_OK != status) {
+                    cl_warn(cl_context->lib,
+                            "tl context %s is not available for sbgp %s",
+                            tls->names[j], ucc_sbgp_str(hs->sbgp_type));
+                } else {
+                    hs->n_tls++;
+                    n_sbgp_teams++;
+                }
+            }
```
Comment (bureddy) [inline src/components/cl/hier/cl_hier_team.c:76]: assert (hs->n_tls <= CL_HIER_MAX_SBGP_TLS) ?

File: src/components/cl/hier/cl_hier_team.c [hunk 12]
```diff
@@ -0,0 +1,305 @@
...
+    }
+
+    j                              = 0;
+    self->team_create_req->n_teams = n_sbgp_teams;
+    /* Initialize base params for ALL tl teams we need to create:
+       for each hier sbgp we have n_tls potentially requested tl teams */
+    for (i = 0; i < UCC_HIER_SBGP_LAST; i++) {
+        hs = &self->sbgps[i];
+        for (t = 0; t < hs->n_tls; t++) {
+            if (hs->state == UCC_HIER_SBGP_ENABLED) {
```
Comment (bureddy) [inline src/components/cl/hier/cl_hier_team.c:92]: swap "if" and "for" statements?

File: src/components/cl/hier/cl_hier_team.c [hunk 13]
```diff
@@ -0,0 +1,305 @@
...
+            }
+        }
+    }
+    status = ucc_team_multiple_req_alloc(&self->team_create_req, n_sbgp_teams);
+    if (UCC_OK != status) {
+        cl_error(cl_context->lib, "failed to allocate team req multiple");
+        goto err;
+    }
+
+    j                              = 0;
```
[the remark this answers is not in the document]
Comment (bureddy) [reply src/components/cl/hier/cl_hier_team.c:91]: sorry, i meant to say `"for (i = 0,  j= 0; "`

**Pull request merged**